### PR TITLE
start roll cheatcode

### PR DESCRIPTION
**Pull request by MaksymilianDemitraszek**

### CODE REVIEW

**Comment by Arcticae** [comment]
@MaksymilianDemitraszek care to take a look?

File: starknet-foundry/crates/forge/tests/data/contracts/constructor_roll_checker.cairo [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+    #[constructor]
+    fn constructor(ref self: ContractState) {
+        let block_numb = starknet::get_block_info().unbox().block_number;
+        self.blk_nb.write(block_numb);
+    }
+
+    #[external(v0)]
+    impl IConstructorRollChecker of super::IConstructorRollChecker<ContractState> {
+        fn was_rolled_on_construction(ref self: ContractState, expected_block_number: u64) {
+            assert(self.blk_nb.read() == expected_block_number, 'block_numb incorrect');
```
Comment (abulenok) [inline starknet-foundry/crates/forge/tests/data/contracts/constructor_roll_checker.cairo:23]: It seems unnatural to have asserts inside the test data. Maybe put it to the test function?

same for  https://github.com/foundry-rs/starknet-foundry/pull/181/files#diff-f77438c3fa2d169290375e9aea86788dfd6b6979080c69d7a8ccadda626c7c8dR18

File: starknet-foundry/crates/cheatable-starknet/src/rpc.rs [hunk 2]
```diff
@@ -0,0 +1,505 @@
...
+        entry_point_selector,
+        calldata,
+        storage_address: contract_address,
+        caller_address: account_address,
+        call_type: CallType::Call,
+        initial_gas: u64::MAX,
+    };
+
+    let mut resources = ExecutionResources::default();
+    let account_context = build_transaction_context();
```
Comment (MaksymilianDemitraszek) [inline starknet-foundry/crates/cheatable-starknet/src/rpc.rs:104]: `build_transaction_context`
always sets nonce to `0`. It is a inconsistent behaviour, can we create an issue to fix it?

File: starknet-foundry/crates/cheatable-starknet/src/rpc.rs [hunk 3]
```diff
@@ -0,0 +1,505 @@
...
+fn execute_if_cairo1(
+    call: CallEntryPoint,
+    contract_class: ContractClass,
+    state: &mut dyn State,
+    cheated_state: &CheatedState,
+    resources: &mut ExecutionResources,
+    context: &mut EntryPointExecutionContext,
+) -> EntryPointExecutionResult<CallInfo> {
+    match contract_class {
+        ContractClass::V0(_) => panic!("Cairo 0 classes are not supported"),
```
[the remark this answers is not in the document]
Comment (Arcticae) [reply starknet-foundry/crates/cheatable-starknet/src/rpc.rs:219]: It will make the old classes behavior a bit unexpected since they cannot be cheated right now - should we consider implementing the cheats for v0 as well?

File: starknet-foundry/crates/cheatable-starknet/src/rpc.rs [hunk 4]
```diff
@@ -0,0 +1,505 @@
...
+}
+// Copied over (with modifications) from blockifier/src/execution/entry_point.rs:144
+fn execute_call_entry_point(
+    entry_point: &mut CallEntryPoint,
+    state: &mut dyn State,
+    cheated_state: &CheatedState,
+    resources: &mut ExecutionResources,
+    context: &mut EntryPointExecutionContext,
+) -> EntryPointExecutionResult<CallInfo> {
+    // Validate contract is deployed.
```
[the remark this answers is not in the document]
Comment (Arcticae) [reply starknet-foundry/crates/cheatable-starknet/src/rpc.rs:154]: Might be hard because of private fields access

File: starknet-foundry/crates/cheatable-starknet/src/rpc.rs (hunk 2 above)
Comment (Arcticae) [reply starknet-foundry/crates/cheatable-starknet/src/rpc.rs:105]: https://github.com/foundry-rs/starknet-foundry/issues/258

File: starknet-foundry/crates/forge/src/running.rs [hunk 5]
```diff
@@ -75,6 +76,9 @@ pub(crate) fn run_from_test_case(
         original_cairo_hint_processor: core_cairo_hint_processor,
         blockifier_state: build_testing_state(predeployed_contracts),
         contracts,
+        cheated_state: CheatedState {
+            rolled_contracts: HashMap::new(),
+        },
```
[the remark this answers is not in the document]
Comment (Arcticae) [reply starknet-foundry/crates/forge/src/running.rs]: [Add a construction metod for CheatedState](https://github.com/foundry-rs/starknet-foundry/pull/181/commits/936c93ebd79ca0e2a729340422bec40b456ced58)

File: starknet-foundry/crates/forge/tests/data/contracts/roll_checker.cairo [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+    #[storage]
+    struct Storage {
+        balance: felt252,
+    }
+
+    #[external(v0)]
+    impl IRollChecker of super::IRollChecker<ContractState> {
+        fn is_rolled(ref self: ContractState, expected_block_number: u64) {
+            let block_numb = starknet::get_block_info().unbox().block_number;
+            assert(block_numb == expected_block_number, 'block_numb incorrect');
```
[the remark this answers is not in the document]
Comment (Arcticae) [reply starknet-foundry/crates/forge/tests/data/contracts/roll_checker.cairo]: [Update logic to getting block number and assertions in tests](https://github.com/foundry-rs/starknet-foundry/pull/181/commits/33421c4864d0e82d0cb0eab28d84410ccc018bb8)

File: starknet-foundry/crates/forge/tests/data/contracts/constructor_roll_checker.cairo (hunk 1 above)
Comment (Arcticae) [reply starknet-foundry/crates/forge/tests/data/contracts/constructor_roll_checker.cairo]: [Update logic to getting block number and assertions in tests](https://github.com/foundry-rs/starknet-foundry/pull/181/commits/33421c4864d0e82d0cb0eab28d84410ccc018bb8)

File: starknet-foundry/crates/cheatable-starknet/src/rpc.rs [hunk 7]
```diff
@@ -0,0 +1,516 @@
...
+            let initial_syscall_ptr = get_ptr_from_res_operand_unchecked(vm, syscall);
+            self.verify_syscall_ptr(initial_syscall_ptr)?;
+
+            // Increment, since the selector was peeked into before
+            self.syscall_handler.syscall_ptr += 1;
+
+            if selector != SyscallSelector::Keccak {
+                self.syscall_handler
+                    .increment_syscall_count_by(&selector, 1);
+            }
```
Comment (cptartur) [inline starknet-foundry/crates/cheatable-starknet/src/rpc.rs:358]: Why we don't increment on Keccak?

File: starknet-foundry/crates/cheatable-starknet/src/rpc.rs [hunk 8]
```diff
@@ -0,0 +1,516 @@
...
+    let args = prepare_call_arguments(
+        &syscall_handler.call,
+        &mut vm,
+        initial_syscall_ptr,
+        &mut syscall_handler.read_only_segments,
+        &entry_point,
+    )?;
+    let n_total_args = args.len();
+
+    // Fix the VM resources, in order to calculate the usage of this run at the end.
```
Comment (cptartur) [inline starknet-foundry/crates/cheatable-starknet/src/rpc.rs:462]: By `fix` you mean `store` right?

File: starknet-foundry/crates/cheatable-starknet/src/rpc.rs
Comment (cptartur) [inline starknet-foundry/crates/cheatable-starknet/src/rpc.rs:1]: It's a good thing we have cheatcodes now, but this code is really hard to follow for me. If we are to get any contributors for this, some refactors should be considered.

File: starknet-foundry/crates/forge/src/cheatcodes_hint_processor.rs [hunk 9]
```diff
@@ -90,6 +88,10 @@ impl ResourceTracker for CairoHintProcessor<'_> {
     }
 }
 
+trait ForgeHintProcessor {
+    fn start_roll(&mut self, inputs: &[Felt252]) -> Result<(), EnhancedHintError>;
```
Comment (cptartur) [inline starknet-foundry/crates/forge/src/cheatcodes_hint_processor.rs:92]: I think it would be better for this to already have properly typed arguments and handle the conversions/extracting values before the call to `start_roll`.

File: starknet-foundry/crates/forge/src/cheatcodes_hint_processor.rs [hunk 10]
```diff
@@ -180,18 +276,36 @@ fn execute_syscall(
 
     let calldata = buffer.next_arr().unwrap();
 
+    let calldata_blockifier: Vec<blockifier_Felt252> = calldata
+        .into_iter()
+        .map(|v| convert_to_blockifier_felt(&v))
+        .collect();
```
Comment (cptartur) [inline starknet-foundry/crates/forge/src/cheatcodes_hint_processor.rs:282]: Not sure, but I think `&v` wouldn't be needed if we used `iter()` instead of `into_iter()`.

File: starknet-foundry/crates/forge/src/cheatcodes_hint_processor.rs [hunk 11]
```diff
@@ -163,10 +248,21 @@ struct ScarbStarknetContractArtifact {
...
 
+fn convert_to_blockifier_felt(val: &Felt252) -> blockifier_Felt252 {
+    let v = val.to_bigint();
+    blockifier_Felt252::from(v)
+}
+
+fn convert_from_blockifier_felt(val: &blockifier_Felt252) -> Felt252 {
+    let v = val.to_bigint();
+    Felt252::from(v)
+}
```
Comment (cptartur) [inline starknet-foundry/crates/forge/src/cheatcodes_hint_processor.rs:259]: This could use some unit tests

File: starknet-foundry/crates/forge/tests/data/contracts/constructor_roll_checker.cairo [hunk 12]
```diff
@@ -0,0 +1,26 @@
...
+        }
+    }
+}
```
Comment (cptartur) [inline starknet-foundry/crates/forge/tests/data/contracts/constructor_roll_checker.cairo:26]: Newlind

File: starknet-foundry/crates/forge/tests/integration/start_roll.rs [hunk 13]
```diff
@@ -0,0 +1,116 @@
...
+        Some(&Utf8PathBuf::from_path_buf(corelib().to_path_buf()).unwrap()),
+        &test.contracts(corelib().path()).unwrap(),
+        &Utf8PathBuf::from_path_buf(predeployed_contracts().to_path_buf()).unwrap(),
+    )
+    .unwrap();
+
+    assert_passed!(result);
+}
+
+// TODO (#254): Make it pass
```
Comment (cptartur) [inline starknet-foundry/crates/forge/tests/integration/start_roll.rs:63]: TODO

File: starknet-foundry/crates/forge/tests/integration/start_roll.rs [hunk 14]
```diff
@@ -0,0 +1,116 @@
+use crate::integration::common::corelib::{corelib, predeployed_contracts};
```
[the remark this answers is not in the document]
Comment (cptartur) [reply starknet-foundry/crates/forge/tests/integration/start_roll.rs:1]: What in the case where we roll A and the A calls other contract B. What should be the expected behavior? 

Should block info in B also be changed?

File: starknet-foundry/crates/forge/tests/integration/start_roll.rs [hunk 15]
```diff
@@ -0,0 +1,116 @@
...
+            use forge_print::PrintTrait;
+            
+            #[starknet::interface]
+            trait IConstructorRollChecker<TContractState> {
+                fn was_rolled_on_construction(ref self: TContractState, expected_block_number: u64);
+            }
+            
+            #[test]
+            fn test_roll_constructor_simple() {
+                assert(1 == 1, 'simple check');
```
Comment (cptartur) [inline starknet-foundry/crates/forge/tests/integration/start_roll.rs:86]: Do we need this?

File: starknet-foundry/crates/forge/tests/integration/start_roll.rs [hunk 16]
```diff
@@ -0,0 +1,116 @@
...
+                let class_hash = declare('ConstructorRollChecker').unwrap();
+                let prepared = PreparedContract { class_hash: class_hash, constructor_calldata: @ArrayTrait::new() };
+                let contract_address: ContractAddress = 2598896470772924212281968896271340780432065735045468431712403008297614014532.try_into().unwrap();
+                start_roll(contract_address, 234);
+                let contract_address: ContractAddress = deploy(prepared).unwrap().try_into().unwrap();
+                contract_address.print();
```
Comment (cptartur) [inline starknet-foundry/crates/forge/tests/integration/start_roll.rs:92]: Is it needed?

File: starknet-foundry/crates/cheatable-starknet/src/rpc.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (MaksymilianDemitraszek) [reply starknet-foundry/crates/cheatable-starknet/src/rpc.rs:229]: We could create a task to triage to think about it

File: starknet-foundry/crates/cheatable-starknet/src/rpc.rs (hunk 7 above)
Comment (MaksymilianDemitraszek) [reply starknet-foundry/crates/cheatable-starknet/src/rpc.rs:358]: Because it has unfixed size

File: starknet-foundry/crates/cheatable-starknet/src/rpc.rs (hunk 8 above)
Comment (MaksymilianDemitraszek) [reply starknet-foundry/crates/cheatable-starknet/src/rpc.rs:462]: Fix, means make it constant at the moment so you can subtract later. Maybe snapshot would be a better wording

File: starknet-foundry/crates/cheatable-starknet/src/rpc.rs [hunk 17]
```diff
@@ -0,0 +1,516 @@
...
+/// Retrieves a [Relocatable] from the VM given a [`ResOperand`].
+/// A [`ResOperand`] represents a CASM result expression, and is deserialized with the hint.
+fn get_ptr_from_res_operand_unchecked(vm: &mut VirtualMachine, res: &ResOperand) -> Relocatable {
+    let (cell, base_offset) = match res {
+        ResOperand::Deref(cell) => (cell, Felt252::from(0)),
+        ResOperand::BinOp(BinOpOperand {
+            op: Operation::Add,
+            a,
+            b: DerefOrImmediate::Immediate(b),
+        }) => (a, Felt252::from(b.clone().value)),
```
[the remark this answers is not in the document]
Comment (MaksymilianDemitraszek) [reply starknet-foundry/crates/cheatable-starknet/src/rpc.rs:288]: It is based on cairo assembly specification. It is cast when we get a pointer incremented by a constant value.

File: starknet-foundry/crates/forge/tests/integration/start_roll.rs (hunk 13 above)
Comment (Arcticae) [reply starknet-foundry/crates/forge/tests/integration/start_roll.rs:63]: it's marked with an issue

File: starknet-foundry/crates/forge/src/cheatcodes_hint_processor.rs (hunk 10 above)
Comment (Arcticae) [reply starknet-foundry/crates/forge/src/cheatcodes_hint_processor.rs:282]: Good catch 👍

File: starknet-foundry/crates/forge/tests/integration/start_roll.rs (hunk 15 above)
Comment (Arcticae) [reply starknet-foundry/crates/forge/tests/integration/start_roll.rs]: [Review](https://github.com/foundry-rs/starknet-foundry/pull/181/commits/26866d978928b80a11135b06c4b672e028872773)

File: starknet-foundry/crates/forge/tests/integration/start_roll.rs (hunk 16 above)
Comment (Arcticae) [reply starknet-foundry/crates/forge/tests/integration/start_roll.rs]: [Review](https://github.com/foundry-rs/starknet-foundry/pull/181/commits/26866d978928b80a11135b06c4b672e028872773)

File: starknet-foundry/crates/cheatable-starknet/src/rpc.rs
Comment (Arcticae) [reply starknet-foundry/crates/cheatable-starknet/src/rpc.rs:1]: Yeah that is a bit of a bummer. The problem that i see that is that we want to keep the code similar to blockifier (i think) and thus more maintainable after bumping, and in the same time keep it neat and clean. We should discuss this, i will create an issue.

Comment (Arcticae) [reply starknet-foundry/crates/cheatable-starknet/src/rpc.rs:1]: https://github.com/foundry-rs/starknet-foundry/issues/264

File: starknet-foundry/crates/cheatable-starknet/src/rpc.rs [hunk 18]
```diff
@@ -0,0 +1,508 @@
...
+        context,
+    )
+    .map_err(|error| {
+        match error {
+            // On VM error, pack the stack trace into the propagated error.
+            EntryPointExecutionError::VirtualMachineExecutionError(error) => {
+                context
+                    .error_stack
+                    .push((storage_address, error.try_to_vm_trace()));
+                // TODO(Dori, 1/5/2023): Call error_trace only in the top call; as it is
```
[the remark this answers is not in the document]
Comment (Arcticae) [reply starknet-foundry/crates/cheatable-starknet/src/rpc.rs:205]: I don't know what you want me to do here exactly

File: starknet-foundry/crates/forge/src/cheatcodes_hint_processor.rs (hunk 10 above)
Comment (Arcticae) [reply starknet-foundry/crates/forge/src/cheatcodes_hint_processor.rs]: [Review 2](https://github.com/foundry-rs/starknet-foundry/pull/181/commits/177b7e3da6761ffb59c65ba1bb3b0f81211af4e2)

File: starknet-foundry/crates/cheatable-starknet/src/rpc.rs (hunk 8 above)
Comment (Arcticae) [reply starknet-foundry/crates/cheatable-starknet/src/rpc.rs]: [Review 2](https://github.com/foundry-rs/starknet-foundry/pull/181/commits/177b7e3da6761ffb59c65ba1bb3b0f81211af4e2)

File: starknet-foundry/crates/cheatable-starknet/src/rpc.rs [hunk 19]
```diff
@@ -0,0 +1,516 @@
...
+fn felt_from_ptr_immutable(
+    vm: &VirtualMachine,
+    ptr: &Relocatable,
+) -> Result<Felt252, VirtualMachineError> {
+    let felt = vm.get_integer(*ptr)?.into_owned();
+    Ok(felt)
+}
+
+impl CheatableSyscallHandler<'_> {
+    fn is_cheated(
```
[the remark this answers is not in the document]
Comment (Arcticae) [reply starknet-foundry/crates/cheatable-starknet/src/rpc.rs]: [Review 2](https://github.com/foundry-rs/starknet-foundry/pull/181/commits/177b7e3da6761ffb59c65ba1bb3b0f81211af4e2)

File: corelib/src/cheatcodes.cairo [hunk 20]
```diff
@@ -81,3 +82,9 @@ fn deploy(prepared_contract: PreparedContract) -> Result::<felt252, RevertedTran
         Result::<felt252, RevertedTransaction>::Err(RevertedTransaction { panic_data })
     }
 }
+
+fn start_roll(contract_address: ContractAddress, block_number: u64) {
+    let contract_address_felt: felt252 = contract_address.into();
+    let block_number_felt: felt252 = block_number.into();
+    cheatcode::<'start_roll'>(array![contract_address_felt, block_number_felt].span());
+}
```
[the remark this answers is not in the document]
Comment (Arcticae) [reply corelib/src/cheatcodes.cairo]: [Review 2](https://github.com/foundry-rs/starknet-foundry/pull/181/commits/177b7e3da6761ffb59c65ba1bb3b0f81211af4e2)

File: starknet-foundry/crates/cheatable-starknet/src/rpc.rs [hunk 21]
```diff
@@ -0,0 +1,508 @@
...
+                StarkFelt::try_from(FAULTY_CLASS_HASH).expect("A class hash must be a felt."),
+            )
+    {
+        return Err(PreExecutionError::FraudAttempt.into());
+    }
+    // Add class hash to the call, that will appear in the output (call info).
+    entry_point.class_hash = Some(class_hash);
+    let contract_class = state.get_compiled_contract_class(&class_hash)?;
+
+    execute_if_cairo1(
```
[the remark this answers is not in the document]
Comment (Arcticae) [reply starknet-foundry/crates/cheatable-starknet/src/rpc.rs]: [Delete obsolete method](https://github.com/foundry-rs/starknet-foundry/pull/181/commits/dd869fd91e9d3259904f5a188d17c06134f981eb)

File: starknet-foundry/crates/forge/tests/integration/start_roll.rs (hunk 14 above)
[the remark this answers is not in the document]
Comment (Arcticae) [reply starknet-foundry/crates/forge/tests/integration/start_roll.rs:1]: @cptartur No

File: starknet-foundry/crates/forge/src/cheatcodes_hint_processor.rs (hunk 9 above)
Comment (Arcticae) [reply starknet-foundry/crates/forge/src/cheatcodes_hint_processor.rs]: https://github.com/foundry-rs/starknet-foundry/pull/181/commits/742c1a1bd03e30f9a1853e5bab6c4bdb9785db02

File: starknet-foundry/crates/forge/src/cheatcodes_hint_processor.rs (hunk 11 above)
Comment (Arcticae) [reply starknet-foundry/crates/forge/src/cheatcodes_hint_processor.rs]: [Add tests for felt conversions](https://github.com/foundry-rs/starknet-foundry/pull/181/commits/ed8bb5f0ccd6217db212addc65080daddeb04dcd)

File: starknet-foundry/crates/forge/tests/integration/start_roll.rs (hunk 14 above)
[the remark this answers is not in the document]
Comment (Arcticae) [reply starknet-foundry/crates/forge/tests/integration/start_roll.rs:1]: [Add a test which tests roll along with other syscall in one function …](https://github.com/foundry-rs/starknet-foundry/pull/181/commits/867415b57d699de462ef6c48d8ec36f5a87d1ea0)

File: starknet-foundry/crates/cheatable-starknet/src/rpc.rs (hunk 17 above)
[the remark this answers is not in the document]
Comment (Arcticae) [reply starknet-foundry/crates/cheatable-starknet/src/rpc.rs:271]: I have no idea TBH xD

[the remark this answers is not in the document]
Comment (Arcticae) [reply starknet-foundry/crates/cheatable-starknet/src/rpc.rs:271]: It's worth hiding this low-level stuff somewhere so it's more understandable when doing #264

File: starknet-foundry/crates/cheatable-starknet/src/rpc.rs [hunk 22]
```diff
@@ -0,0 +1,516 @@
...
+        references: &[HintReference],
+    ) -> Result<Box<dyn Any>, VirtualMachineError> {
+        self.syscall_handler
+            .compile_hint(hint_code, ap_tracking_data, reference_ids, references)
+    }
+}
+
+/// Retrieves a [Relocatable] from the VM given a [`ResOperand`].
+/// A [`ResOperand`] represents a CASM result expression, and is deserialized with the hint.
+fn get_ptr_from_res_operand_unchecked(vm: &mut VirtualMachine, res: &ResOperand) -> Relocatable {
```
[the remark this answers is not in the document]
Comment (Arcticae) [reply starknet-foundry/crates/cheatable-starknet/src/rpc.rs:264]: Why not?

File: starknet-foundry/crates/cheatable-starknet/src/rpc.rs [hunk 23]
```diff
@@ -0,0 +1,499 @@
...
+        &args,
+        &mut run_resources,
+        verify_secure,
+        Some(program_segment_size),
+        vm,
+        hint_processor,
+    )?;
+
+    Ok(run_resources)
+}
```
Comment (piotmag769) [inline starknet-foundry/crates/cheatable-starknet/src/rpc.rs:499]: Was comparing with blockifier, why do we care only about run resources?

File: starknet-foundry/crates/cheatable-starknet/src/rpc.rs [hunk 24]
```diff
@@ -0,0 +1,499 @@
...
+        &mut self,
+        vm: &mut VirtualMachine,
+        exec_scopes: &mut ExecutionScopes,
+        hint_data: &Box<dyn Any>,
+        constants: &HashMap<String, Felt252>,
+        run_resources: &mut RunResources,
+    ) -> HintExecutionResult {
+        let maybe_extended_hint = hint_data.downcast_ref::<Hint>();
+
+        if let Some(Hint::Starknet(StarknetHint::SystemCall { system: _ })) = maybe_extended_hint {
```
Comment (piotmag769) [inline starknet-foundry/crates/cheatable-starknet/src/rpc.rs:240]:
```suggestion
        if let Some(Hint::Starknet(StarknetHint::SystemCall { .. })) = maybe_extended_hint {
```

**Review by piotmag769 — COMMENTED** [review]
I think it would be highly beneficial if we could have comments stating which functions are copied 1:1 from external sources. If a function is partially modified we could also inform about it (even like this: // MODIFIED  // END_MODIFIED). Later on we will forget and it will become unmaintainable. 

It would also makes reviews much easier. For example I don't know what parts of `execute_next_syscall_cheated` are ours and what should I focus on during a review.

File: starknet-foundry/crates/cheatable-starknet/src/rpc.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (MaksymilianDemitraszek) [reply starknet-foundry/crates/cheatable-starknet/src/rpc.rs:164]: We could have our own field to track it or sth

File: starknet-foundry/crates/forge/src/cheatcodes_hint_processor.rs [hunk 25]
```diff
@@ -163,10 +258,21 @@ struct ScarbStarknetContractArtifact {
...
```
[the remark this answers is not in the document]
Comment (Arcticae) [reply starknet-foundry/crates/forge/src/cheatcodes_hint_processor.rs]: that's actually pretty nice idea.

File: starknet-foundry/crates/forge/src/cheatcodes_hint_processor.rs [hunk 26]
```diff
@@ -163,10 +258,21 @@ struct ScarbStarknetContractArtifact {
...
+
 fn execute_syscall(
     system: &ResOperand,
     vm: &mut VirtualMachine,
     blockifier_state: &mut CachedState<DictStateReader>,
+    cheated_state: &CheatedState,
```
[the remark this answers is not in the document]
Comment (Arcticae) [reply starknet-foundry/crates/forge/src/cheatcodes_hint_processor.rs]: We will address such things while refactoring, resolving

File: starknet-foundry/crates/forge/tests/data/contracts/roll_checker.cairo [hunk 27]
```diff
@@ -0,0 +1,21 @@
+#[starknet::interface]
```
[the remark this answers is not in the document]
Comment (Arcticae) [reply starknet-foundry/crates/forge/tests/data/contracts/roll_checker.cairo:1]: https://github.com/foundry-rs/starknet-foundry/issues/279

File: starknet-foundry/crates/cheatable-starknet/src/rpc.rs (hunk 23 above)
Comment (Arcticae) [reply starknet-foundry/crates/cheatable-starknet/src/rpc.rs:499]: What do you mean? I return them here for compatibility, since i can't pass mutable reference into that function

Comment (piotmag769) [reply starknet-foundry/crates/cheatable-starknet/src/rpc.rs:504]: In blockifier something else is returned I think, so I was wondering if we are not losing any data here

File: starknet-foundry/crates/forge/src/cheatcodes_hint_processor.rs (hunk 25 above)
[the remark this answers is not in the document]
Comment (Arcticae) [reply starknet-foundry/crates/forge/src/cheatcodes_hint_processor.rs]: Not doable since this crate does not own this type

**Review by piotmag769 — APPROVED** [review]

File: starknet-foundry/crates/forge/src/cheatcodes_hint_processor.rs (hunk 26 above)
[the remark this answers is not in the document]
Comment (MaksymilianDemitraszek) [reply starknet-foundry/crates/forge/src/cheatcodes_hint_processor.rs]: @piotmag769 Misleading name imo as it executes all hints

**Review by MaksymilianDemitraszek — COMMENTED** [review]
Apporve ✅ (my PR)

File: starknet-foundry/crates/forge/src/cheatcodes_hint_processor.rs [hunk 28]
```diff
@@ -747,3 +713,65 @@ mod test {
...
+            BLOCKIFIER_FIELD_HIGH,
+        ] {
+            let from = cairo_Felt252::from(value);
+            let to = convert_to_blockifier_felt(&from);
+            assert_eq!(
+                to.to_u128().expect("Downcast failure"),
+                from.to_u128().expect("Downcast failure")
+            );
+        }
+    }
```
Comment (cptartur) [inline starknet-foundry/crates/forge/src/cheatcodes_hint_processor.rs:758]: Can't we use some hardcoded expected values? Converting to and from and comparing the values seems like it could not detect some issues.

Comment (Arcticae) [reply starknet-foundry/crates/forge/src/cheatcodes_hint_processor.rs:758]: It is about checking numerical correctness of the conversion, so it seems alright to me.

Comment (MaksymilianDemitraszek) [reply starknet-foundry/crates/forge/src/cheatcodes_hint_processor.rs:758]: Agree with @Arcticae here

Comment (cptartur) [reply starknet-foundry/crates/forge/src/cheatcodes_hint_processor.rs:758]: Not a fan of this approach

**Review by cptartur — APPROVED** [review]
